src/common/utils/math.ts

```diff
@@ -0,0 +1,96 @@
+const mod = (val: number, n: number) => {
+  return ((val % n) + n) % n;
+};
+
+/**
+ * Returns the smallest power of BASE greater than or equal to N
+ * @param base number the base of the logarithm
+ * @param n number the number to take the log of
+ * @returns number
+ */
+const powRoundUp = (base: number, n: number) => {
+  const ceilLog = Math.ceil(Math.log(n) / Math.log(base));
+  return Math.pow(base, ceilLog);
+};
+
+/**
+ * Uses the Box-Muller transform to generate Gaussian samples
+ * @param mean
+ * @param variance
+ * @returns number
+ */
+const gaussian = (mean: number = 0, variance: number = 1) => {
+  var u1 = 0,
+    u2 = 0;
+  // Math.random() returns a value in [0, 1) but the B-M transform requires values in (0, 1)
+  // Need to make sure neither value is 0
+  while (u1 === 0) u1 = Math.random();
+  while (u2 === 0) u2 = Math.random();
+  const R = Math.sqrt(-2 * Math.log(u1));
+  const theta = 2 * Math.PI * u2;
+  const stdNormal = R * Math.cos(theta);
+  return (stdNormal + mean) * variance;
+};
+
+/**
+ * Initializes an empty 2-D array
+ * @param size
+ * @returns
+ */
+const initArray = (size: number) => {
+  let array: Array<Array<number>> = new Array(size);
+  for (let i = 0; i < size; i++) {
+    array[i] = new Array(size);
+  }
+  return array;
+};
+
+/**
+ * Passes a radius 1 box blur (3x3 kernel) over an array
+ * in-place using the fact that the box blur is a separable
+ * filter by passing a moving average filter horizontally
+ * then vertically. This construction has O(1) memory and
+ * O(N) time where N is the number of entries in the array
+ * @param array2D
+ * @returns Array<Array<number>>
+ */
+const boxBlur = (array2D: Array<Array<number>>) => {
+  const wrapArray = (idx: number) => {
+    return mod(idx, array2D.length);
+  };
+  var left, mid, right, top, bot;
+
+  // Apply 1-D moving average to each row
+  for (var i = 0; i < array2D.length; i++) {
+    left = 0;
+    mid = array2D[i][wrapArray(-1)] / 3;
+    right = array2D[i][0] / 3;
+    for (var j = 0; j < array2D.length; j++) {
+      left = mid;
+      mid = right;
+      right = array2D[i][wrapArray(j + 1)] / 3;
+      array2D[i][j] = left + mid + right;
+    }
+  }
+
+  // Apply 1-D moving average to each column
+  for (j = 0; j < array2D.length; j++) {
+    top = 0;
+    mid = array2D[wrapArray(-1)][j] / 3;
+    bot = array2D[0][j] / 3;
+    for (i = 0; i < array2D.length; i++) {
+      top = mid;
+      mid = bot;
+      bot = array2D[wrapArray(i + 1)][j] / 3;
+      array2D[i][j] = top + mid + bot;
+    }
+  }
+};
```

Comment:
You can just combine this into one double for loop.

```suggestion
  for (var i = 0; i < array2D.length; i++) {
	// For rows
    left = 0;
    right = array2D[i][0] / 3;
    mid_row = array2D[i][wrapArray(-1)] / 3;

	// For cols
    top = 0;
    bot = array2D[0][j] / 3;
    mid_col = array2D[wrapArray(-1)][j] / 3;

    for (var j = 0; j < array2D.length; j++) {
  	  // Apply 1-D moving average to each row
      left = mid;
      mid_row = right;
      right = array2D[i][wrapArray(j + 1)] / 3;
      array2D[i][j] = left + mid + right;

	  // Apply 1-D moving average to each column
	  top = mid;
      mid_col = bot;
      bot = array2D[wrapArray(j + 1)][i] / 3;
      array2D[j][i] = top + mid + bot;
    }
  }
```

Reply:
will look into this later, but are you sure this is correct? I'm not sure that this actually replicates behavior; for example, if we have a 3x3:
```
1 2 3
4 5 6
7 8 9
```
iterating through the for-loop once would yield:
```
(1+2+4) 2 3
4       5 6
7       8 9
```
when in reality we want the first entry to be `(1 + 2 + 4 + 5)`?

Reply:
No, I think you're right --  they're not the same. I think the intended behavior is for the rows to be computed first, then you perform the column-wise computation on the result. Whereas when you combine it into one double for loop, you're doing the col-wise computation before the rows have been properly computed. I'm sure you can still put this into one double for loop, but it may take much more time, and not worth the effort.

For context, these are the results with my proposed changes: https://jsfiddle.net/nmdtqe0k/2/, and these are the results from what you originally had: https://jsfiddle.net/wepfdx2g/4/